### DISCUSSION TITLE
If I want to write an YAML generator for Crossplane, what can I use?

### DISCUSSION THREAD

**Discussion by danielpsf** [crossplane/crossplane#2272] [Q&A]
Well, we're building our IDP at the moment and one of the challenges we're facing is that developers currently don't want to deal with Terraform and most likely won't like to deal with Crossplane definition files as well (not all, we have 350+ devs at the moment, so you can imagine it is a diverse population). So, one of our plan to facilitate the adoption is to create a CLI capable of abstract their needs, at least in the beginning, to wrap their requests in CLI commands that would, in the end, generate files on their disk that they can commit to Github and move through the pipeline.

Not going to the merit of Crossplane being amazingly documented, but you know, sometimes it is better to teach them how to do something simple using a CLI (i.e: `myawesomecliname cloud-provider create database --type SQL` and voila) other than having them doing all sorts of requests for our infrastructure/platform team to do this files for them or blindly copying and pasting stuff from other places.

Well, in the end, my question is: "where can I find the structs so that we don't have to re-invent the wheel?"
- Can we rely on the Crossplane providers and use their structs as defined like https://github.com/crossplane/provider-aws/blob/master/apis/database/v1beta1/rdsinstance_types.go? Or would it be too weak of a solution?

**Comment by grantgumina** [comment]
I'd love to learn more about what you imagine your developer's workflows being. Are you using [Composition](https://crossplane.io/docs/v1.1/getting-started/create-configuration.html) yet? One thing we see organizations using Crossplane doing is putting all of their configuration in a repo, and having branches for each team which include templated YAML files for claims. Those teams then use the templates to provision new resources by putting the claim through a pipeline which verifies and actually applies the changes to a Crossplane instance.

**Reply by halcyondude** [reply]
A CLI based approach to supporting interactive generation of artifacts/assets from live infra has a lot of potential. 

I am exploring patterns / methods to import existing mission-critical infrastructure investments. Empowering users to quickly  demonstrate the value of crossplane in their existing settings has many downstream benefits.  It should be easy, clear and deterministic to generate Compositions / XRD's w/ definitions for existing infrastructure. 

Providing this capability helps advocates and adopters to demonstrate the value of the meta-control plane approach to _**their**_ stakeholders in familiar terms. Doing so without requiring up-front, non-trivial time investment for Implementors and/or Operators increases the viability of presenting & recommending this type of solution to Platform Owners or descision-makers.

It will enable transfer of ownership --> crossplane "For the Staging cluster(s) as a Pilot." Tooling tends to be most useful when it's not a "bolt on" - but rather is merely using the in-tree dev tools. Examples include:

* [Angular's](https://angular.io/) "[ng generate](https://angular.io/cli/generate)" and it's use of schematics
* [.NET](https://dotnet.microsoft.com)'s "[dotnet new](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-new)"

These support interactive workflows that grant **agency** to users. Because they can selectively generate (and re-generate) the artifacts, the user is engaged and building confidence as they learn how Compositions and XRD are crafted - again using their own infra as the subject - to interactively generate the requisite artifacts/files. This folds nicely into whatever source control & editors are in use. 

They could take as input cloud-specific JSON, incorporating helpers to fetch it directly from aws/gcp/azure/tanzu/..., using existing credentials/profiles/mechanisms already present on the user's machine. This  approach enables straightforward generation of test collateral as a perk.

The implementation of the CLI ("generators" ?) could employ a compiler-style approach.

* **_front end_**:  compile cloud-specific JSON (the existing infra) to an intermediate representation / IL suitable to pass to a back end that emits artifacts.  Define plugin interface, optionally implemented in-tree with providers, leveraging a utility library. This ensures plugins don't drift, as they can be CI'd w/ the providers.
* **_back_end_**: IL --> Composition/XRD's + .yaml artifacts. Provided by crossplane for all providers. In the AWS example the plugin would generate Crossplane collateral leveraging the [ACK](https://github.com/aws-controllers-k8s)

## Use Case: Atlassian Standard Infrastructure (ASI) on AWS

Plucked from the wild today.

ASI -> https://aws.amazon.com/quickstart/architecture/atlassian-standard-infrastructure

* Deploy ASI (via CF)
  * Generate opinionated Compositions/XRD, test collateral, best practice project, and definitions matching what's been deployed.
  * Experiment with assuming ownership of resources.
  * Experiment with provisioning net-new to a new account

* create new resources in the account (bucket, db, api gateway, ...), via whatever existing collateral...exists.
  * Generate
  * commit, gitops, assume ownership of that too

* Deploy other Atlassian Applications atop ASI, generating new `kind:` of things along the way.

Finally, and perhaps in my view most strategically, this approach to DX allows cloud providers to continue to differentiate on their own UI's, while not demanding that Adopters stop using these UI's. 

Existing ansible/puppet/chef/bash/CDK/terraform/jsonnet/mouseclick/... becomes a modelling tool :)

WDYT?  How much of this is possible today, what are the gaps?
